Refactor: Extend GitHub class as GitHubAnnos, add tests

pbinkley opened the pull request:
This PR extracts the github utility functions from flaskserver.py and incorporates them into a new class GitHubAnnos, which extends GitHub (from github-flask). There's not much to test here (since we're not going to test the GitHub API) but I've added a couple of tests of createdatadict to show that it is making the correct choices based on input.

dnoneill commented on annonatate/utils/github.py:
```diff
@@ -0,0 +1,53 @@
+
+class GitHubAnno(GitHub):
+  def setbork(self):
+    self.bork = 'bork'
```
What is this? Are these necessary lines?

pbinkley replied:
sorry, just diagnostics while I was making sure the new GitHubAnno class was being used. I've deleted them in my second commit.

dnoneill commented on annonatate/flaskserver.py:
```diff
@@ -35,7 +35,7 @@
                   )
 Session(app)
-github = GitHub(app)
+github = GitHubAnno(app)
```
I am very confused by this. It is working but I am not sure why it is.

pbinkley replied:
GitHubAnno is a new class which inherits from the GitHub class defined in flask-github. It can do everything that the GitHub  class could do, plus it has newly-defined methods. This allows encapsulating all of annonatate's github communication in a single class.

dnoneill replied:
That is what I thought but wasn't sure how it "knew" we were passing the app but I am not that much an expert on Inheritance.

dnoneill commented on annonatate/utils/github.py:
```diff
@@ -0,0 +1,53 @@
+          data['sha'] = sha
+      return {'data':data, 'url':full_url}
+
+  def updateAnnos(self, session):
```
moving this out is making it not useful in terms of functionality. The main reason it is there is to update the annotations faster than the api. It is updating the session['annotations'], something it can not do in this utils file.

pbinkley replied:
Hmm, I didn't see that. I'll have another look. It could move back into the main app.

pbinkley replied:
I think this should be ok: updateAnnos updates the ```session``` that is passed to it, which would normally be the main session (e.g. in flaskserver.py at ```githubresponse = github.updateAnnos(session)```. I'm just adding a test to confirm this.

dnoneill replied:
updateAnnos is used twice. It causes problems where it does not have an equal.

dnoneill replied:
I am thinking I might want to get rid of this. I am not sure how useful it is. It was put in place to solve two problems. If two people are working on the application at the same time during the period where we are not querying the API, it would grab any annotations from the GitHub repo so both could see what the other is doing. 
The other is to remove any deleted annotations that did not get deleted in API because it is still rebuilding. I would love to hear your thoughts.

pbinkley replied:
I'd vote for keeping it - it doesn't look fragile, and while it might not come into play very often, when it does it might protect someone from having their data messed up. I haven't thought about live collaborative annotating, but I can imagine it would be a cool way for work for some projects.

pbinkley commented on annonatate/flaskserver.py:
```diff
@@ -704,16 +704,6 @@ def parsetype(board):
-    try:
-        githubresponse = github.get(session['currentworkspace']['contents_url'].replace('{+path}', filepath))
-        githubfilenames = list(map(lambda x: x['name'], githubresponse))
-        session['annotations'] = list(filter(lambda x: x['filename'].split('/')[-1] in githubfilenames,annotations))
```
@dnoneill Another clarification request: why does this line use ```x['filename'].split('/')[-1]``` instead of just ```x['name']```? I'm finding that ```githubresponse``` is in the same annotation format as what's stored in ```session['annotations']```, so there are no filename attriibutes, but ```x['name']``` matches the name in ```githubfilenames```. Are there circumstances where this wouldn't be true? I've got a test now that uses ```x['name'] and shows that the local deletion of an annotation that is now missing on Github works.

dnoneill replied:
Because we are comparing the annotations in the session, which don't have a key that is name, they just have the full url which is under the key filename. While the GitHub files we found, which have the key name are in the format where they are just the filename ('xxxx-list.json', 'xxxxx.json') so we need to the the portion that we can match on. I am not quite sure what that test is showing but session annotations never have the key "name". You can see that in the API, https://dnoneill.github.io/annonatate/, which has the fields: filename, json and order.

pbinkley replied:
OK, thanks - I need to fix my test data. I'm giving it github annotations instead of session annotations.

dnoneill replied:
Ah. That makes sense. Glad that was semi coherent. I was worried you might not understand what I was talking about because it seemed rambly to me.

dnoneill commented on annonatate/flaskserver.py:
```diff
@@ -704,16 +704,6 @@ def parsetype(board):
-def updateAnnosGitHub():
-    annotations = session['annotations']
-    try:
-        githubresponse = github.get(session['currentworkspace']['contents_url'].replace('{+path}', filepath))
```
filepath is in the settings.py file. I am actually removing this with the wax integration but you should probably put it in.

https://github.com/dnoneill/annonatate-sourcecode/blob/main/annonatate/settings.py#L5

pbinkley replied:
Of course! I should have thought to look there.

pbinkley commented:
This new commit adds a test for githubAnnos. To keep the test simple, I'm passing "hybrid" annotations (with both ```name``` and ```filename``` attributes), so that githubAnnos can use them as github annotations and as schema annotations at the same time. This seems ok to me since what we're testing is just the ability to extract the list of filenames and match them up to ```session['annotations']```; but if that seems kludgy and you'd rather have it keep the two separate I'm happy to revise. Note that the test also shows that the session that is passed in has been updated when it is tested at the end.

pbinkley commented on test/test_github.py:
```diff
@@ -60,6 +60,26 @@ def test_search_query_upload(self):
+    keep_annotation = { "name": "annotation_to_keep", "filename": "http://test/annotation_to_keep" }
+    delete_annotation =   { "name": "annotation_to_delete", "filename": "http://test/annotation_to_delete" }
+
+    mock_raw_request = MagicMock()
```
drat, I meant to change the name ```mock_raw_request``` to ```mock_github_get```. I'll do that along with whatever other changes are still needed.

dnoneill commented:
@pbinkley The session updates in the utils file, however it does not update the session in the flaskserver. This is the real problem. You pass session as a variable, however when you are doing anything with session in the utils file, it is acting as a variable not as a part of the session. It updates in the function, but you are not updating it in the flaskserver which means it doesn't really update. Does that make sense?

pbinkley commented:
@dnoneill I was hoping that ```test_updateAnnos``` showed that the session variable that gets passed to ```updateAnnos``` comes back modified (since it goes in with two annotations but then is tested to contain 1), and that the same would be true of the flaskserver session which gets passed in at [line 634](https://github.com/pbinkley/annonatate-sourcecode/blob/4e1a7a827ad40099c9926d2591f6e117a8ca70b5/annonatate/flaskserver.py#L634) (and similarly with the other GitHubAnnos methods). Are you sure that it isn't? I wasn't sure of the best way to test this in real time.

dnoneill commented:
@pbinkley Pretty sure you would have to set the line to session['annotations'] = github.updateAnnos(session, filepath) where the updateAnnos would return something. The best way to test this would be to run the application. 
change this line from
if 'annotations' not in session.keys() or session['annotations'] == '' or  duration > 35:
to 
if 'annotations' not in session.keys() or session['annotations'] == '' or  duration > 1:

Then create an annotation, delete an annotation and then go to the /annotations page.
If it is working correctly the annotation that has been deleted should not show up on the annotations page. If it isn't working correctly it will keep showing up until the site rebuilds.

pbinkley commented:
I'll try that. Changing what updateAnnos returns reminds me: currently it always return false (due to the ```fnally``` clause. and in ```getAnnotations()``` there's [a ```if githubresponse``` block](https://github.com/pbinkley/annonatate-sourcecode/blob/4e1a7a827ad40099c9926d2591f6e117a8ca70b5/annonatate/flaskserver.py#L641-L659) that will therefore never be executed sinice ```githubresponse``` is always false. Can I safely delete this block?

dnoneill commented:
@pbinkley in response to (it is not showing up for me): I'll try that. Changing what updateAnnos returns reminds me: currently it always return false (due to the fnally clause. and in getAnnotations() there's a if githubresponse block that will therefore never be executed sinice githubresponse is always false. Can I safely delete this block?


No. If it is always returning False there is a major bug in the code. It should not be returning False all the time. I think you need to get rid of `finally`, and use `except` instead. Finally is always run even if the code is executing properly, which means the return statement at the end of the code is probably not firing. That finally means the code is failing and there is a problem somewhere. That bit of code you want to delete is the bit that allows multiple users to be working on the same repository at the same time with very little lag time.

pbinkley commented:
Sorry, I deleted the comment because I realized I don't know whether it
really does return false from the finally clause if it has already returned
githubresponse at the end of the try clause (assuming no exception was
thrown).

Peter Binkley, Ph.D., MLIS / Digital Scholarship Technologies Librarian /
***@***.***
2-10K Cameron Library / University of Alberta / Edmonton, Alberta / Canada
T6G 2J8
phone 780-492-3743 / fax 780-492-9243
he/him/his

*The University of Alberta respectfully acknowledges that we are situated
on Treaty 6 territory, traditional lands of First Nations and Métis people.*


On Fri, 6 Aug 2021 at 16:24, dnoneill ***@***.***> wrote:

> @pbinkley <https://github.com/pbinkley> in response to (it is not showing
> up for me): I'll try that. Changing what updateAnnos returns reminds me:
> currently it always return false (due to the fnally clause. and in
> getAnnotations() there's a if githubresponse block that will therefore
> never be executed sinice githubresponse is always false. Can I safely
> delete this block?
>
> No. If it is always returning False there is a major bug in the code. It
> should not be returning False all the time. I think you need to get rid of
> finally, and use except instead. Finally is always run even if the code
> is executing properly, which means the return statement at the end of the
> code is probably not firing. That finally means the code is failing and
> there is a problem somewhere. That bit of code you want to delete is the
> bit that allows multiple users to be working on the same repository at the
> same time with very little lag time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dnoneill/annonatate-sourcecode/pull/2#issuecomment-894547723>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACCTMAI3V4ZI7GILBSJXH3T3ROKLANCNFSM5BPY46IA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

pbinkley commented:
@dnoneill I've tried the test you suggested, and it looks like session['annotations'] is being updated. I report the annotation count before and after calling ```updateAnnos()``` in ```getannotations()```:

```
print("Annotations before: " + str(len(session['annotations'])))
github.updateAnnos(session, filepath)
print("Annotations after: " + str(len(session['annotations'])))
```
Then I deleted an annotation (by deleting the file in [my _annotations folder](https://github.com/pbinkley/annonatate/tree/main/_annotations) in the Github web interface), and immediately refreshed the [flaskserver annotations page](http://0.0.0.0:5000/annotations/). The result was:

```
Annotations before: 10
Annotations after: 9
```

I like the idea of returning session['annotations'] and assigning it explicitly, but that raised the question of what happens if I change what updateAnnos returns, which prompted me to think about the finally clause.

pbinkley commented:
@dnoneill I've confirmed that the finally was indeed causing it to return False always. I've change it to except and it now gets an array of annotations in ```githubresponse````, so the block runs.

pbinkley commented:
@dnoneill I've committed the change from finally to except, and I've left in my diagnostics to show what updateAnnos is doing to flaskserver's session. Would you care to try it and make sure it's behaving the way it should? It looks ok to me but I dont have all the logic clear in my mind.

dnoneill commented:
@pbinkley There was a bug from when I switched Annonatate from v2 to v3. I think I fixed everything in the new commits. Please take a look and see if I missed anything. If not I think this is ready to merge.

pbinkley commented:
@dnoneill I think it's good to merge. I've modified the test slightly to separate the session and github versions of my test annotations, so that someone reading the test won't be fooled into thinking all annotations are **supposed** to have a ```name``` and a ```filename``` attribute, and I've removed my diagnostics.
